text/0000-hidden-impls.md

```diff
@@ -0,0 +1,669 @@
+
+As Rust's trait system (and thus implementations on types thereof) has its
+direct descendant in Haskell, we consider how that language deals with
+the relation between modules and type class instances (same as `impl` in Rust).
+While Haskell - without certain extensions enabled, like Rust,
+[guarantees coherence] of type class instances, it does not guarantee
+global uniqueness of instances. Haskell will allow you to define overlapping
+instances but won't allow you to actually use them - producing an error once
+you try to do that.
+
+[optional_doi]: https://link.springer.com/chapter/10.1007/978-3-319-45279-1_9
+[optional_pdf]: http://homepages.dcc.ufmg.br/~camarao/CT/optional-type-classes.pdf
+
+In a paper [Optional Type Classes for Haskell][optional_pdf] due to
+Ribero et. al [(DOI: 10.1007/978-3-319-45279-1_9)][optional_doi]
+modularization of instances is considered. This is however radically different
+than what is proposed in this RFC, which does *not* propose modularization of
+implementations. Fundamentally, that paper seems permit different instances
+of the same type class to coexist as long as the instance to use for a given
+expression is uniquely determinable. As previously discussed, such a mechanism
+would not be sound in Rust. The main difference between these approaches is
+that in our proposal, all implementations are considered when coherence checking
+an implementation. The effect of that is to prevent different crates and modules
+from defining implementations for a trait and a type when a hidden one exists.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Exact wordings of error messages discussed in this RFC and their error codes,
+as well as lint names and their phrasings are out of the scope of this RFC. 
+
+The following questions are in the scope of the RFC and should be resolved
+*before* the RFC is merged:
+
+1. Are there any strange interactions with possible plans for negative bounds?
+
+2. Should `crate` (and `pub(crate)`) be the only visibility modifier permitted
+   on `impl`?
+
+3. Should `priv` be introduced to hide implementations from other modules?
+
+4. Should you be able to specify a visibility modifier in `#[derive(..)]`?
+
+   1. Should `#[structural_match]` take visibility of `impl`s into account?
+
+   2. Should the visibility modifier be automatically added to all
+      trait implementations emitted by a custom derive macro?
+
+The following questions should be resolved at least before stabilization
+(but possibly sooner):
+
+5. What is an appropriate default lint level for `hidden_fn`?
+   1. Should `hidden_fn` be promoted to `deny`-by-default?
+   2. should it be `allow`-by-default?
+   3. Should it perhaps be a hard error instead?
+
+6. Should `pub impl ..` be linted against suggesting `impl ..` instead?
+
+7. Should `priv field: Type` and the `priv` visibility modifier in general
+   be linted against when it is the contextual default?
```

Comment:
I'd like to say that the `priv` vs `pub(self)` etc debate is also something to solve *after* merging the RFC.

Reply:
Absolutely; Now that I know `pub(self)` is stable, we should punt on `priv`.